### ISSUE TITLE
how can we combine two pulses?

### ISSUE THREAD

**Issue by kuangshupei** [laserfun/laserfun#77]
If I have created a pulse1 at wavelength1 and a pulse2 at wavelength2, how to combine the two into a new pulse instance?

**Comment by DanHickstein** [comment]
There isn't currently a built-in function to do this, but, it's straightforward to do. We just need to shift the spectrum of one of the pulses onto the grid of the other pulse. Here is some example code:

```python
import numpy as np
import matplotlib.pyplot as plt
import scipy
import laserfun as lf

# create two pulses
pulse0 = lf.Pulse(pulse_type='sech', fwhm_ps=0.1, center_wavelength_nm=1550, time_window_ps=10, epp=1e-9)         
pulse1 = lf.Pulse(pulse_type='sech', fwhm_ps=0.1, center_wavelength_nm=1750, time_window_ps=10, epp=1e-9)

# create a third pulse that starts the same as pulse 0, but will have pulse 1 added to it later
pulse2 = pulse0.create_cloned_pulse()

# re-interpolate the spectrum of pulse1 onto the grid of pulse0
interpolator = scipy.interpolate.interp1d(pulse1.f_THz, pulse1.aw, bounds_error=False, fill_value=0.0)
new_aw = interpolator(pulse0.f_THz)

# combine the pulses:
pulse2.aw = (pulse0.aw + new_aw)

# plot the results:
fig, axs = plt.subplots(3, 1, figsize=(5, 8), tight_layout=True)

axs[0].plot(pulse0.f_THz, np.abs(pulse0.aw)**2)
axs[1].plot(pulse1.f_THz, np.abs(pulse1.aw)**2)
axs[2].plot(pulse2.f_THz, np.abs(pulse2.aw)**2)

for ax in axs:
    ax.grid(alpha=0.2)
    ax.set_xlim(130, 230)

axs[-1].set_xlabel('Frequency (THz)')
axs[-1].set_ylabel('Intensity')

plt.show()
```

This could be an easy function to add to laserfun. What is the best way to do it? Perhaps a function in the Pulse class that is called "add_pulses", which could be called like:

```python
pulse0.add_pulse(pulse1)
```

This could either modify pulse0 or return a new pulse object. I'm not sure which is better.

**Comment by kuangshupei** [comment]
Thanks, I get your point, by doing this, pulse0, pulse1, and pulse2 all have the same frequency grid right? I also had a look at the PyNLO's interpolate_to_new_center_wl function, which basically use the same interpolate method, however, the documentation of this function says "Change grids by interpolating the electric field onto a new frequency grid, defined by the new center wavelength and the current pulse parameters. This is useful when grid overlaps must be avoided, for example in difference or sum frequency generation", in such wave-mixing processes, shouldn't we use different grids according to this comment?

**Comment by kuangshupei** [comment]
> There isn't currently a built-in function to do this, but, it's straightforward to do. We just need to shift the spectrum of one of the pulses onto the grid of the other pulse. Here is some example code:
> 
> ```python
> import numpy as np
> import matplotlib.pyplot as plt
> import scipy
> import laserfun as lf
> 
> # create two pulses
> pulse0 = lf.Pulse(pulse_type='sech', fwhm_ps=0.1, center_wavelength_nm=1550, time_window_ps=10, epp=1e-9)         
> pulse1 = lf.Pulse(pulse_type='sech', fwhm_ps=0.1, center_wavelength_nm=1750, time_window_ps=10, epp=1e-9)
> 
> # create a third pulse that starts the same as pulse 0, but will have pulse 1 added to it later
> pulse2 = pulse0.create_cloned_pulse()
> 
> # re-interpolate the spectrum of pulse1 onto the grid of pulse0
> interpolator = scipy.interpolate.interp1d(pulse1.f_THz, pulse1.aw, bounds_error=False, fill_value=0.0)
> new_aw = interpolator(pulse0.f_THz)
> 
> # combine the pulses:
> pulse2.aw = (pulse0.aw + new_aw)
> 
> # plot the results:
> fig, axs = plt.subplots(3, 1, figsize=(5, 8), tight_layout=True)
> 
> axs[0].plot(pulse0.f_THz, np.abs(pulse0.aw)**2)
> axs[1].plot(pulse1.f_THz, np.abs(pulse1.aw)**2)
> axs[2].plot(pulse2.f_THz, np.abs(pulse2.aw)**2)
> 
> for ax in axs:
>     ax.grid(alpha=0.2)
>     ax.set_xlim(130, 230)
> 
> axs[-1].set_xlabel('Frequency (THz)')
> axs[-1].set_ylabel('Intensity')
> 
> plt.show()
> ```
> 
> This could be an easy function to add to laserfun. What is the best way to do it? Perhaps a function in the Pulse class that is called "add_pulses", which could be called like:
> 
> ```python
> pulse0.add_pulse(pulse1)
> ```
> 
> This could either modify pulse0 or return a new pulse object. I'm not sure which is better.

Thanks, I get your point, by doing this, pulse0, pulse1, and pulse2 all have the same frequency grid right? I also had a look at the PyNLO's interpolate_to_new_center_wl function, which basically use the same interpolate method, however, the documentation of this function says "Change grids by interpolating the electric field onto a new frequency grid, defined by the new center wavelength and the current pulse parameters. This is useful when grid overlaps must be avoided, for example in difference or sum frequency generation", in such wave-mixing processes, shouldn't we use different grids according to this comment?

**Comment by DanHickstein** [comment]
Ah, good eye! I did borrow the interpolation approach from PyNLO :).

But no, `pulse0` and `pulse1` are on different grids. The `center_wavelength_nm=` argument determines the grid (along with `time_window_ps` and `npts`). We need to re-interpolate the shape of `pulse1` to put it on the same grid as `pulse0`.

I also wondered about that comment in PyNLO. Perhaps @ycasg can elaborate if he still remembers. In addition to simulating Chi3 nonlinear optics with the NLSE, PyNLO also had functions for simulating chi2 processes like sum and difference frequency generation. My guess is that certain gridding situations could cause artifacts for the chi2 processes. I don't know if this logic applies to the NLSE.

What pulse wavelengths are you trying to simulate? If you're using pulses that are very separated in frequency, then your temporal waveform will likely have rapid modulations. My understanding is that this kinda violates the "slowly varying envelope" approximation that the NLSE makes. So, the results may be somewhat unpredictable. That said, I have previously used the NLSE to simulate the interaction of a pulse at 1 micron with a pulse at 0.5 micron, and the results seemed to match my experiment reasonably well, so who knows! :)

**Comment by kuangshupei** [comment]
Right! You are correct about the pulses having different grids in your approach mentioned above. I have been using PyNLO and using the method like `pulse0 = pynlo.light.DerivedPulses.SechPulse(pulseWL + other parameters here)` and `pulse1 = pynlo.light.DerivedPulses.SechPulse(pulseWL/2 + parameters here), pulse1 = pulse1.interpolate_to_new_center_wl(pulseWL)` before accessing their spectral field amplitude AW. This makes them already appear on the same grid before combination. But they both turn out to be equivalent ;)

**Comment by kuangshupei** [comment]
Btw I am using pulseWL=1035nm and pulseWL/2=517.5nm. I also found that `pulse0 = pynlo.light.DerivedPulses.SechPulse(pulseWL + other parameters here)` and `pulse1 = pynlo.light.DerivedPulses.SechPulse(pulseWL + parameters here), pulse1.interpolate_to_new_center_wl(pulseWL/2)` still works, but by doing so they are on different grids, and the second harmonic pulse still has center wavelength of pulseWL/2. The representation depends on which grid and which centre wavelength we use (when plotting spectral intensity w.r.t. frequency or when using the spectrogram function). And the two pulses are in each other's frequency position due to interpolation, so the energy allocation needs to be reversed. This method matches the comment in PyNLO but is quite counterintuitive.

**Comment by DanHickstein** [comment]
Sounds good. I'm slightly confused about your last comment, but if it seems to be working for you, that's great!
